pkg/tarmak/logs/logs.go

```diff
@@ -0,0 +1,337 @@
+	Gid                     string `json:"_GID"`
+	Comm                    string `json:"_COMM"`
+	Exe                     string `json:"_EXE"`
+	Cmdline                 string `json:"_CMDLINE"`
+	SystemdCGroup           string `json:"_SYSTEMD_CGROUP"`
+	SystemdSession          string `json:"_SYSTEMD_SESSION"`
+	SystemdOwnerUID         string `json:"_SYSTEMD_OWNER_UID"`
+	SystemdUnit             string `json:"_SYSTEMD_UNIT"`
+	SourceRealtimeTimestamp string `json:"_SOURCE_REALTIME_TIMESTAMP"`
+	MachineID               string `json:"_MACHINE_ID"`
+	Hostname                string `json:"_HOSTNAME"`
+
+	// We require interface here since Message is not always a string. In the
+	// case of vault-assets.service for example, some output is displayed as a
+	// byte slice.
+	Message interface{} `json:"MESSAGE"`
+}
+
+func New(tarmak interfaces.Tarmak) *Logs {
+	return &Logs{
+		tarmak: tarmak,
+		log:    tarmak.Log(),
+		ctx:    tarmak.CancellationContext(),
+	}
+}
+
+func (l *Logs) Gather(group string, flags tarmakv1alpha1.ClusterLogsFlags) error {
+	l.log.Infof("fetching logs from target '%s'", group)
+	if err := l.initialise(group, flags); err != nil {
+		return fmt.Errorf("failed to set tar gz log bundle path: %v", err)
+	}
+
+	err := l.ssh.WriteConfig(l.tarmak.Cluster())
+	if err != nil {
+		return err
+	}
+
+	select {
+	case <-l.ctx.Done():
+		return l.ctx.Err()
+	default:
+	}
+
+	dir, err := ioutil.TempDir("", filepath.Base(l.path))
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(dir)
+	l.tmpDir = dir
+
+	var result *multierror.Error
+	hostGatherFunc := func(host string) {
+		defer l.wg.Done()
+
+		reader, writer := io.Pipe()
+		go l.readStream(reader, host, result)
+
+		l.log.Infof("fetching journald logs from instance '%s'", host)
+		err := l.fetchCmdOutput(
+			host,
+			"journalctl",
+			[]string{"-o", "json", "--no-pager", "--since", l.since, "--until", l.until},
+			writer,
+		)
+		if err != nil {
+			l.mu.Lock()
+			result = multierror.Append(result, err)
+			l.mu.Unlock()
+		}
+	}
+
+	aliases, err := l.hostAliases()
+	if err != nil {
+		return err
+	}
+
+	for _, a := range aliases {
+		if err := os.Mkdir(
+			filepath.Join(l.tmpDir, a),
+			os.FileMode(0755)); err != nil {
+			return err
+		}
+
+		l.wg.Add(1)
+		go hostGatherFunc(a)
+	}
+
+	l.wg.Wait()
+
+	if result != nil {
+		return result.ErrorOrNil()
+	}
+
+	select {
+	case <-l.ctx.Done():
+		return l.ctx.Err()
+	default:
+	}
+
+	for _, f := range l.tmpFiles {
+		f.Close()
+	}
+
+	return l.bundleLogs()
+}
+
+func (l *Logs) readStream(reader io.Reader, host string, result *multierror.Error) {
+	readerScanner := bufio.NewScanner(reader)
+
+	for readerScanner.Scan() {
+		r := readerScanner.Bytes()
+
+		if r == nil || len(r) == 0 {
+			continue
+		}
+
+		entry := new(SystemdEntry)
+		if err := json.Unmarshal(r, entry); err != nil {
+			l.mu.Lock()
+			err = fmt.Errorf("failed to unmarshal entry [%s]: %s", r, err)
+			result = multierror.Append(result, err)
+			l.mu.Unlock()
```

Comment:
Keep the are with the lock separate from other code (new lines before and after). also keep it minimal (i.e.) without the Errorf